fontir/src/variations.rs

```diff
@@ -0,0 +1,795 @@
+///
+/// This algorithm started in MutatorMath and was subsequently modified in FontTools.
+///
+/// This is a Rust version of FontTools getMasterLocationsSortKeyFunc.
+/// <https://github.com/fonttools/fonttools/blob/2f1f5e5e7be331d960a0e30d537c2b4c70d89285/Lib/fontTools/varLib/models.py#L295>
+struct LocationSortingHat {
+    axis_order: Vec<String>,
+    on_axis_points: HashMap<String, HashSet<NormalizedCoord>>,
+}
+
+impl LocationSortingHat {
+    fn new(locations: &[NormalizedLocation], axis_order: Vec<String>) -> LocationSortingHat {
+        // Location is on-axis if it has exactly 1 non-zero coordinate
+        let mut on_axis_points: HashMap<String, HashSet<NormalizedCoord>> = HashMap::new();
+        'location: for location in locations {
+            let mut on_axis: Option<(&String, NormalizedCoord)> = None;
+            for (axis_name, pos) in location.iter() {
+                if pos.into_inner() == ZERO {
+                    continue;
+                }
+                if on_axis.is_some() {
+                    continue 'location; // multiple non-zero coords, bail out
+                }
+                on_axis = Some((axis_name, *pos));
+            }
+            if let Some((axis_name, pos)) = on_axis {
+                on_axis_points
+                    .entry(axis_name.clone())
+                    .or_default()
+                    .insert(pos);
+            }
+        }
+        LocationSortingHat {
+            axis_order,
+            on_axis_points,
+        }
+    }
+
+    fn key_for(&self, location: &NormalizedLocation) -> LocationSortKey {
+        let mut rank = 0;
+        let mut on_axis_points: i16 = 0;
+        let mut ordered_axes = Vec::new();
+        let mut non_zero_axes = Vec::new();
+
+        // Score the index of the axis for any axis with non-zero value in the user specified axis order
+        let mut known_axes = Vec::new();
+        for (idx, axis_name) in self.axis_order.iter().enumerate() {
+            if location.has_non_zero(axis_name) {
+                known_axes.push(idx);
+                ordered_axes.push(axis_name.clone());
+            }
+        }
+
+        for (axis_name, pos) in location.iter() {
+            if pos.into_inner() != ZERO {
+                // Rank is the # of axes with non-zero values
+                rank += 1;
+
+                // Score more than the max allowable axes for any axis with non-zero value that has no user-specified order
+                if !self.axis_order.contains(axis_name) {
+                    known_axes.push(0x10000);
+                }
+
+                non_zero_axes.push(axis_name);
+            }
+            // score -1 for every axis position that matches a location that was directly on an axis
+            on_axis_points += self
+                .on_axis_points
+                .get(axis_name)
+                .map(|on_axis| on_axis.get(pos).map_or(0, |_| -1))
+                .unwrap_or_default();
```

Comment:
honestly this chain of method calls looks a bit dense/convoluted.. you want to count the number of on-axis points for the given location, and negate that number so the sort is decreasing (`-len(onPointAxes)`). Here you decrease a counter -1 for each value that is an on-axis point, am I right?

EDIT: Ok nevermind, I read the docs, over time I will get more familiar with all these methods. All good, nothing to do here.